### ISSUE TITLE
Status of C/C++ APIs?

### ISSUE THREAD

**Issue by matteosal** [tensorflow/tensorflow#85958]
Do they fully support running operators and models, gradients and optimizers? What's the relationship between them? Are there up-to-date resources, tutorials etc on how to use them?

**Comment by Venkat6871** [comment]
Hi **@matteosal** ,
Welcome to TensorFlow! Yes, TensorFlow fully supports running models, computing gradients, and using optimizers. These are fundamental building blocks for building and training machine learning models in TensorFlow. Here, I am providing the [documentation](https://www.tensorflow.org/js/guide/models_and_layers) for your reference. If you have any further queries, please fill out all the required templates.

Thank you!

**Comment by matteosal** [comment]
@Venkat6871 you have linked a page about Javascript, I asked for C/C++. I am asking this because it's hard to find clear information about them. 

[This document](https://github.com/tensorflow/docs/blob/master/site/en/r1/guide/extend/bindings.md) states at the bottom that the C API does not support gradients, but maybe that's for TF 1 so maybe it doesn't apply anymore.

Then I found this [C++ API reference page](https://www.tensorflow.org/api_docs/cc/) which just lists a bunch of symbols with no contextual explanation or introduction. Looking around there only seem to be some isolated gradient primitives but I couldn't find anything analogous to a generic method of running backpropagation through an arbitrary set of operators.

**Comment by matteosal** [comment]
@mihaimaruseac 

> The C API is very limited, it's used to create bindings to other languages.

Then it seems natural to expect it supports gradients, right? It's a core feature anyone wanting to create a binding for another language would need.

**Comment by mihaimaruseac** [comment]
[This](https://github.com/tensorflow/tensorflow/blob/40e02fe076f822840e3d52c95d40613317aeb5de/tensorflow/c/c_api.h) is the entire C API. In particular [it supports gradients](https://github.com/tensorflow/tensorflow/blob/40e02fe076f822840e3d52c95d40613317aeb5de/tensorflow/c/c_api.h#L994).

**Comment by matteosal** [comment]
@mihaimaruseac 
Yes I noticed that function in the header. It has a [note](https://github.com/tensorflow/tensorflow/blob/40e02fe076f822840e3d52c95d40613317aeb5de/tensorflow/c/c_api.h#L990-L993) which states that not all gradients are supported. The link is broken but it's clear it's meant to lead [here](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/cc/gradients). This document states that implemented gradients should be declared as `<OpName>Grad(...)`, and looking for this template in the entire directory produces these matches:
```
[Abs, Acos, Acosh, Add, AddN, Angle, Asin, Asinh, Atan, Atan2, Atanh, BaseFusedBatchNorm, BatchMatMul, BatchMatMulV2, BatchToSpace, BatchToSpaceND, BroadcastTo, Cast, CheckNumerics, ClipByValue, Complex, ConcatV2, Conj, Conv2D, Conv2DBackpropInput, Cos, Cosh, Cumsum, DepthToSpace, DepthwiseConv2dNative, Diag, DiagPart, Div, DivNoNan, DynamicPartition, DynamicStitch, Einsum, Erf, Erfinv, Exp, ExpandDims, Expm1, Fill, FusedBatchNormV3, GatherNd, GatherV2, Identity, Imag, Inv, L2Loss, Lgamma, Log, Log1p, LogSoftmax, MatMul, MatrixBandPart, MatrixDiag, Maximum, Mean, Minimum, MinOrMax, MirrorPad, MirrorPadGrad, Mul, Ndtri, Neg, Pack, Pad, PartitionedCall, Pow, Prod, QuantizeAndDequantize, QuantizeAndDequantizeV3, ReadVariableOp, Real, RealDiv, RefIdentity, Reshape, Reverse, ReverseSequence, Roll, Rsqrt, ScatterNd, ScatterNdNonAliasingAdd, SegmentSum, Select, SelectV2, Sigmoid, Sign, Sin, Sinh, Slice, Softmax, SoftmaxCrossEntropyWithLogits, SpaceToBatch, SpaceToBatchND, SpaceToDepth, Split, SplitV, Sqrt, Square, SquaredDifference, Squeeze, Sub, Sum, Tan, Tanh, Tile, Transpose, Unpack, UnsortedSegmentMinOrMax, UnsortedSegmentSum]
```

Several things are missing here, e.g. 1D and 3D Convolutions, ConvTranspose, Pooling, spatial resampling, RNNs. Can anyone confirm this is correct and the above gradients are the only ones available?

**Comment by mihaimaruseac** [comment]
That is correct. PRs welcome to add more support if needed, though I think JAX is a better replacement.

**Comment by matteosal** [comment]
Thank you
